### ISSUE TITLE
[Bug] Silent fp32 GEMM corruption on gfx1201 when M > 2^19 rows (both rocBLAS and hipBLASLt; torch 2.13.0+rocm7.2)

### ISSUE THREAD

**Issue by bioritmovideo** [ROCm/ROCm#6595]
## Problem description

On **gfx1201** (Radeon AI PRO R9700), fp32 GEMMs (`torch.mm` / `addmm` / `bmm` /
`F.linear`) with **more than 2^19 = 524,288 rows** return silently corrupted results.
**No error or warning is raised** — the op "succeeds" with wrong math.

The two BLAS paths fail differently (both verified element-wise against a float64 CPU
reference on a 200k-row random sample):

- **hipBLASLt path** (`TORCH_BLAS_PREFER_HIPBLASLT=1`): corruption is surgical — rows up to
  ~524,287 are correct, **every row from ~524,288 on is garbage** (≈ bias-only values).
- **rocBLAS path** (`TORCH_BLAS_PREFER_HIPBLASLT=0`), `addmm` with bias: once M exceeds the
  threshold, **nearly the whole result is wrong** (bad rows starting from row 1 —
  199,998/200,000 sampled rows wrong at M=1.68M).
- `M = 524,280` and `M = 524,288` are both fully correct on both paths; `M = 600,000` breaks.
- **fp16 / bf16 at the same sizes are correct** (tested up to M = 1.68M).
- The 2^19 boundary suggests an unhandled grid-dimension limit in the large-M SGEMM
  kernel launch/tiling path.

## Environment

| | |
|---|---|
| GPU | 2× AMD Radeon AI PRO R9700 (gfx1201, RDNA4) |
| ROCm | 7.2.1 |
| PyTorch | 2.13.0+rocm7.2 (official wheel) |
| OS | Ubuntu 24.04.4 LTS, kernel 6.17.0-35-generic |

## Reproducer and actual output

Reproducer (self-contained, ~45 lines, runs in ~30 s): [repro_sgemm_fp32_gfx1201.py](https://github.com/bioritmovideo/trellis2-rocm-gfx1201/blob/master/repro_sgemm_fp32_gfx1201.py)
Actual output on our hardware:

```
=== rocBLAS path (TORCH_BLAS_PREFER_HIPBLASLT=0) ===
[PROBE] M=1676291 K=64 N=7 bias=True: bad_rows=199998/200000 min_bad_row=1 max_good_row=951899 biasonly_dist=0.0204 absmean_bad=0.7329
[PROBE] M=524280 K=64 N=7 bias=True: bad_rows=0/200000
[PROBE] M=524288 K=64 N=7 bias=True: bad_rows=0/200000
[PROBE] M=600000 K=64 N=7 bias=True: bad_rows=50479/200000 min_bad_row=2 max_good_row=553795 biasonly_dist=0.0329 absmean_bad=0.5287
[PROBE] M=1048576 K=64 N=7 bias=True: bad_rows=199996/200000 min_bad_row=19 max_good_row=445586 biasonly_dist=0.0505 absmean_bad=1.0006
[PROBE] M=1676291 K=1728 N=64 bias=False: bad_rows=137692/200000 min_bad_row=524291 max_good_row=524284 absmean_bad=0.7605

=== hipBLASLt path (TORCH_BLAS_PREFER_HIPBLASLT=1) ===
[PROBE] M=1676291 K=64 N=7 bias=True: bad_rows=137445/200000 min_bad_row=524295 max_good_row=524286 biasonly_dist=1.1386 absmean_bad=0.8482
[PROBE] M=524280 K=64 N=7 bias=True: bad_rows=0/200000
[PROBE] M=524288 K=64 N=7 bias=True: bad_rows=0/200000
[PROBE] M=600000 K=64 N=7 bias=True: bad_rows=25198/200000 min_bad_row=524288 max_good_row=524283 biasonly_dist=0.9897 absmean_bad=0.7831
[PROBE] M=1048576 K=64 N=7 bias=True: bad_rows=100040/200000 min_bad_row=524291 max_good_row=524283 biasonly_dist=1.2262 absmean_bad=0.6807
[PROBE] M=1676291 K=1728 N=64 bias=False: bad_rows=137692/200000 min_bad_row=524291 max_good_row=524284 absmean_bad=0.7605
```

Note the `min_bad_row ≈ 524,288` boundary on the hipBLASLt path and on the bias-free
shape, versus `min_bad_row=1` on the rocBLAS `addmm` path.

## Real-world impact

Found while porting **Microsoft TRELLIS.2** (image-to-3D) to gfx1201: at 1024³ voxel
resolution the sparse-VAE decoder runs its final LayerNorm+Linear in fp32 over ~1.68M
voxel rows → most occupancy logits came back as bias values → silently shredded meshes
(exit 0, statistics of intermediate tensors looked normal). The same bug hit `torch.bmm`
over 1.68M vertices in the mesh renderer. At 512³ (~405k rows, under the threshold)
everything is correct, which made this hard to track down.

Workaround we ship: chunk the matmul into ≤262,144-row blocks (mathematically identical).

This looks like the same *class* of silent gfx12 numerical bug as #5981 (Hunyuan3D texture
corruption on R9700, fixed via TheRock nightlies in Feb 2026).

## Ask

- Confirm and route to the right component (rocBLAS / hipBLASLt / Tensile kernel selection).
- Happy to test TheRock nightlies or candidate fixes on this hardware — the reproducer
  takes ~30 seconds.
